src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,433 @@
+                    }
+                }
+            };
+
+        final ArchCondition<JavaClass> beAnnotatedOrSuppressed = new ArchCondition<>(
+            "be annotated or added to the suppression list") {
+            @Override
+            public void check(JavaClass item, ConditionEvents events) {
+                if (!SUPPRESSED_CLASSES.contains(item.getFullName())) {
+                    final String message = String
+                        .format(Locale.ROOT, "should %s.", getDescription());
+                    events.add(SimpleConditionEvent.violated(item, message));
+                }
+            }
+        };
+
+        final ArchCondition<JavaField> beSuppressed = new ArchCondition<>("be suppressed") {
+            @Override
+            public void check(JavaField item, ConditionEvents events) {
+                if (!SUPPRESSED_FIELDS_IN_MODULES.contains(item.getFullName())) {
+                    final String message = String
+                        .format(Locale.ROOT, "should %s.", getDescription());
+                    events.add(SimpleConditionEvent.violated(item, message));
+                }
+            }
+        };
+
+        final ArchRule fieldsInStatelessChecksAreImmutable = fields()
+            .that()
+            .areDeclaredInClassesThat()
+            .areAnnotatedWith(StatelessCheck.class)
+            .and(are(not(moduleProperties)))
+            .should(BE_IMMUTABLE)
+            .andShould()
+            .beFinal()
+            .andShould(bePrivateExceptMessageKeys)
+            .orShould(beSuppressed);
+
+        final ArchRule classesWithImmutableFieldsShouldBeStateless = classes()
+            .that(have(immutableFields))
+            .and()
+            .haveSimpleNameEndingWith("Check")
+            .and()
+            .doNotHaveModifier(JavaModifier.ABSTRACT)
+            .should()
+            .beAnnotatedWith(StatelessCheck.class)
+            .orShould(beAnnotatedOrSuppressed);
+
+        final ArchRule classesWithMutableFieldsShouldBeStateful = classes()
+            .that(doNot(have(immutableFields)))
+            .and()
+            .haveSimpleNameEndingWith("Check")
+            .and()
+            .doNotHaveModifier(JavaModifier.ABSTRACT)
+            .should()
+            .beAnnotatedWith(FileStatefulCheck.class)
+            .orShould()
+            .beAnnotatedWith(GlobalStatefulCheck.class)
+            .orShould(beAnnotatedOrSuppressed);
+
+        fieldsInStatelessChecksAreImmutable.check(checkstyleChecks);
+        classesWithImmutableFieldsShouldBeStateless.check(checkstyleChecks);
+        classesWithMutableFieldsShouldBeStateful.check(checkstyleChecks);
+    }
+
+    private static final class ImmutableFieldArchCondition extends ArchCondition<JavaField> {
+        private ImmutableFieldArchCondition() {
+            super("be among immutable types");
+        }
+
+        private static boolean isMutable(JavaField javaField) {
```

Comment:
I feel like checking for `isMutable` in `ImmutableFieldArchCondition` is strange... It feels like we should be checking for immutability if anything, or tweak class to be `MutableFieldArchcondition` and negate it.

Reply:
Done.